plugins/schema/graphql-kotlin-sdl-generator/src/main/kotlin/com/expediagroup/graphql/plugin/schema/GenerateSDL.kt

```diff
@@ -0,0 +1,110 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.graphql.plugin.schema
+
+import com.expediagroup.graphql.SchemaGeneratorConfig
+import com.expediagroup.graphql.TopLevelObject
+import com.expediagroup.graphql.extensions.print
+import com.expediagroup.graphql.federation.FederatedSchemaGeneratorConfig
+import com.expediagroup.graphql.federation.FederatedSchemaGeneratorHooks
+import com.expediagroup.graphql.federation.toFederatedSchema
+import com.expediagroup.graphql.hooks.NoopSchemaGeneratorHooks
+import com.expediagroup.graphql.server.operations.Mutation
+import com.expediagroup.graphql.server.operations.Query
+import com.expediagroup.graphql.server.operations.Subscription
+import com.expediagroup.graphql.plugin.schema.hooks.SchemaGeneratorHooksProvider
+import com.expediagroup.graphql.toSchema
+import io.github.classgraph.ClassGraph
+import io.github.classgraph.ScanResult
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.util.ServiceLoader
+import kotlin.RuntimeException
+
+private val logger: Logger = LoggerFactory.getLogger("generateSDL")
+
+/**
+ * Generates GraphQL schema in SDL format.
+ *
+ * Relies on a `ServiceLoader` mechanism to dynamically load an instance of a `SchemaGeneratorHooksProvider` from the classpath.
+ */
+fun generateSDL(supportedPackages: List<String>): String {
+    val hooksProviders = ServiceLoader.load(SchemaGeneratorHooksProvider::class.java).toList()
+    val hooks = when {
+        hooksProviders.isEmpty() -> {
+            logger.warn("No SchemaGeneratorHooksProvider were found, defaulting to NoopSchemaGeneratorHooks")
+            NoopSchemaGeneratorHooks
+        }
+        hooksProviders.size > 1 -> {
+            throw RuntimeException("Cannot generate SDL as multiple SchemaGeneratorHooksProviders were found on the classpath")
+        }
+        else -> {
+            val provider = hooksProviders.first()
+            logger.debug("SchemaGeneratorHooksProvider found, ${provider.javaClass.simpleName} will be used to generate the hooks")
+            provider.hooks()
+        }
+    }
+    val scanResult = ClassGraph()
+        .enableAllInfo()
+        .acceptPackages(*supportedPackages.toTypedArray())
+        .scan()
+
+    val queries = findTopLevelObjects(scanResult, Query::class.java)
+    val mutations = findTopLevelObjects(scanResult, Mutation::class.java)
+    val subscriptions = findTopLevelObjects(scanResult, Subscription::class.java)
+
+    // TODO support top level name overrides?
```

Comment:
We support the name overrides in the generator but I am wondering if we should just drop support for those all together. It should be common practice now to have the query object just called `Query`

Reply:
The alternative is that we move the top level name override to the hooks but that doesn't seem like the best approach to me

Reply:
Thats why i did not implement this yet :) - we can pass those values from plugin task/mojo directly. Technically renaming those top fields is valid per spec but yeah unsure how often this is used.